src/_pytest/assertion/rewrite.py

```diff
@@ -302,21 +302,25 @@ def _write_pyc_fp(
     # as of now, bytecode header expects 32-bit numbers for size and mtime (#4903)
     mtime = int(source_stat.st_mtime) & 0xFFFFFFFF
     size = source_stat.st_size & 0xFFFFFFFF
+    # 64-bit source file hash
```

Comment:
My understanding from PEP 552 is that for if we want to use checked-hash pyc files, we need to set the lowest bit of the `flags` to indicate it's a hash-based, and the 2nd lowest bit to indicate that it's checked, and get rid of the mtime and size bytes.

Reply:
Hmm, I specifically what to use all three (`size`, `mtime`, and `hash`) to get the best performance. That's how other applications to do it as well (e.g. black, mypy). It will thus deviate from CPython but I guess the comment here is relevant
https://github.com/pytest-dev/pytest/blob/71f265f1f32c7d9884b07f08819fb148b5b13521/src/_pytest/assertion/rewrite.py#L303-L305

Reply:
You are correct, but I wonder if we can reduce complexity by just always doing the hashing.

Were you able to get a measurable benefit in home-assistant repo by using this branch?

Are you using the `--check-hash-based-pycs` Python flag in home-assistant? If so, with which value?

Reply:
> Were you able to get a measurable benefit in home-assistant repo by using this branch?

Not yet. It's on my todo list. Although it's sadly not as easy as adding caching to the CI job. We use `pytest-xdist` with 10 different jobs, so I might have to combine the different cache files later based on mtimes. That's something I'll explore soon.

> You are correct, but I wonder if we can reduce complexity by just always doing the hashing.

That's certainly an option, however I don't think it's really that complicated. The combined logic is quite simple. Even if calculating the hash is fast, we still need to read the source file first to calculate it. That's not necessary if we are sure that it isn't a match `-> different size`, or that it is because the `mtimes` are the same.

> Are you using the `--check-hash-based-pycs` Python flag in home-assistant? If so, with which value?

I'll probably need to use `always` for Home Assistant, just because the mtimes are reset when the cache is restored., but that's only for the CI environment. I won't use that locally! _(The biggest speedup from not always using the hash will also be locally.)_

Reply:
I like that the pytest pyc files are actually pyc files. It allows inspecting them using standard tools and such. We can definitely diverge, then we should change the magic and file extension to make it clear, but I'd rather not if possible.

I wish Python just added the hash on top of the mtime/size instead of replacing them, but what can you do. So I think it will be best to take on the little bit of extra complexity and implement the check-hash-based-pycs logic. The value is available in `_imp.check_hash_based_pycs` (since `_imp` is private, we should fall back to `default` if the import fails). The cpython logic is in [here](https://github.com/python/cpython/blob/75cd86599bad05cb372aed9fccc3ff884cd38b70/Lib/importlib/_bootstrap_external.py#L1071).

Assuming that caching CIs that want the hashing behavior would want to set `check-hash-based-pycs` anyway, while local users don't, that would do the right thing.

Reply:
Sounds reasonable. Especially the part that following the pyc file layout allows inspecting them using standard tools. Let me see what I can do.

Reply:
I've been looking at the source code and reading [PEP 552](https://peps.python.org/pep-0552/). Unless I'm missing something `--check-hash-based-pycs` doesn't actually affect how pycs files are written. Even with it set to `always`, it won't overwrite timestamp based pycs files. The only way to generate hashed based ones seems to be using `python -m compile` or `python -m compileall` with either `--invalidation-mode checked-hash` or `SOURCE_DATE_EPOCH` set.

This certainly complicates things a bit for caching the pycs files in CI jobs as I'll likely need to add a precompile job.

However, the relevant question here now is, how do we want to handle it in pytest? Would it make sense to add a `--invalidation-mode` flag with `timestamp | checked-hash` values? `--check-hash-based-pycs` won't be that useful probably as it seems `default` will always be enough.

Reply:
Hmm you are correct, I assumed that `--check-hash-based-pycs` also affects the writing but it doesn't!

The `--invalidation-mode` flag is starting to get a little complicated... Before we proceed, can you please try your branch with `--invalidation-mode checked-hash` and check that it makes a measurable impact?

If we decide that it's actually worth it, I hope we can just use the checked-hash mode always. I wrote a script to check the hashing overhead:

```py
import time, pathlib
from importlib.util import source_hash

hashing_time, count, size = 0, 0, 0
for entry in pathlib.Path('/usr/lib/python3.11').glob("**/*.py"):
    source = entry.read_bytes()
    count += 1
    size += len(source)
    start = time.perf_counter()
    source_hash(source)
    hashing_time += time.perf_counter() - start
print(f'{count} files, {size/1024/1024} MiB, {hashing_time} seconds')
```

On my lousy 2015 laptop, it gives

```
17699 files, 273.7363386154175 MiB, 0.10418461405697599 seconds
```

This seems fast enough to me to not worry about. (BTW, the current PR code does the hashing always even in timestamp mode, so the overhead is present anyway, but that is fixable).

Reply:
> The `--invalidation-mode` flag is starting to get a little complicated... Before we proceed, can you please try your branch with `--invalidation-mode checked-hash` and check that it makes a measurable impact?

Took some time but I got it working now. That impact is smaller than I thought but still measurable. Including the time it takes to pre-populate the cache with (`pytest --invalidation-mode checked-hash --collect-only`), it's about a 6% improvement.

The 6% number is for caching only the pytest pyc files. Haven't gotten to it yet, but I also like to check if caching the regular Python pyc files helps as well. Need to pre-compile them manually though as Python would always write timestamp based pyc files.

Another idea, although specific to the Home Assistant setup, would be to explore how that can be better used with `pytest-test-groups`. If we already run the `--collect-only` job, maybe collection can be skipped entirely for the actual pytest runs?

Reply:
10% is a good improvement. Do you happen to have a measurement of the initial (uncached) hash-computing run vs. a timestamp one? This is to measure the hashing overhead, if any.

As I said in my previous comment, I think we can just use hash mode always and it will be fine.

> Another idea, although specific to the Home Assistant setup, would be to explore how that can be better used with pytest-test-groups. If we already run the --collect-only job, maybe collection can be skipped entirely for the actual pytest runs?

Caching collection is a thorny issue, I'd avoid attempting it unless you are willing to put in a lot of effort.